Git-commit: 3233661c817bcf3024cfe5d960736f5b6d7cbbb1
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Fri, 26 Feb 2021 20:21:03 +0800
Subject: [PATCH] arm64/mpam: operation not permitted when remove a ctrl group
 with a mondata
Patch-mainline: Queued in openEuler repo, version 5.10.0-4.5.0
References: openEuler-21.03
Git-repo: https://gitee.com/openeuler/kernel.git

hulk inclusion
category: feature
feature: ARM MPAM support
bugzilla: 48265
CVE: NA

--------------------------------

root@(none)$ ls
cpus        cpus_list   ctrlmon     mon_data    mon_groups  schemata    tasks
root@(none)$ cd ../
root@(none)$ rmdir p2
rmdir: 'p2': Operation not permitted

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Kai Liu <kai.liu@suse.com>
---
 fs/resctrlfs.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/fs/resctrlfs.c b/fs/resctrlfs.c
index 39bbb6e1fd88..255b91623392 100644
--- a/fs/resctrlfs.c
+++ b/fs/resctrlfs.c
@@ -208,7 +208,7 @@ static int resctrl_group_create_info_dir(struct kernfs_node *parent_kn)
 	}
 
 	/*
-	 * This extra ref will be put in kernfs_remove() and guarantees
+	 m This extra ref will be put in kernfs_remove() and guarantees
 	 * that @rdtgrp->kn is always accessible.
 	 */
 	kernfs_get(kn_info);
@@ -329,6 +329,7 @@ static int resctrl_get_tree(struct fs_context *fc)
 
 		kernfs_get(kn_mongrp);
 
+#ifndef CONFIG_ARM64 /* [FIXME] arch specific code */
 		ret = mkdir_mondata_all(resctrl_group_default.kn,
 					&resctrl_group_default, &kn_mondata);
 		if (ret)
@@ -336,6 +337,7 @@ static int resctrl_get_tree(struct fs_context *fc)
 
 		kernfs_get(kn_mondata);
 		resctrl_group_default.mon.mon_data_kn = kn_mondata;
+#endif
 	}
 
 	ret = kernfs_get_tree(fc);
@@ -347,8 +349,10 @@ static int resctrl_get_tree(struct fs_context *fc)
 	goto out;
 
 out_mondata:
+#ifndef CONFIG_ARM64 /* [FIXME] arch specific code */
 	if (resctrl_mon_capable)
 		kernfs_remove(kn_mondata);
+#endif
 out_mongrp:
 	if (resctrl_mon_capable)
 		kernfs_remove(kn_mongrp);
@@ -847,6 +851,11 @@ static void resctrl_group_rm_ctrl(struct resctrl_group *rdtgrp, cpumask_var_t tm
 static int resctrl_group_rmdir_ctrl(struct kernfs_node *kn, struct resctrl_group *rdtgrp,
 			       cpumask_var_t tmpmask)
 {
+#ifdef CONFIG_ARM64 /* [FIXME] arch specific code */
+	if (rdtgrp->flags & RDT_CTRLMON)
+		return -EPERM;
+#endif
+
 	resctrl_group_rm_ctrl(rdtgrp, tmpmask);
 
 	/*
-- 
2.31.1

